### ISSUE TITLE
Not able to run late_fusion.m

### ISSUE THREAD

**Issue by christofjan** [bryanyzhu/Hidden-Two-Stream#21]
Hi Sir,
              I am new to ubuntu and am not able to run late_fusion.m file with the command 
math -script late_fusion.m. how did you run this file. 
Thank you..

**Comment by bryanyzhu** [comment]
Hi, .m file is matlab file. You need to have matlab installed to run it. 

If you don't have access to matlab, you can rewrite the late_fusion.m using python. Basically, it is just a weighted averaging procedure, should be easy to write.

**Comment by christofjan** [comment]
thanks for the information.

**Comment by christofjan** [comment]
Hi sir,
           I installed matlab and i tried to run the file late_fusion.m file but i'm getting following error. could you please suggest me how to resolve this. thank you.

Error using strsplit (line 80)
First input must be either a character vector or a string scalar.

Error in late_fusion (line 15)
    tmp{i} = strsplit(groudTruth{i}, '/');

**Comment by bryanyzhu** [comment]
This is some error in Matlab, I didn't come across this before, and I can't reproduce it on my end. So I don't know how to solve it for you. 

But it should be easy, the error log is pretty clear. Just print out to see what are those variables. Maybe the matlab version, or maybe the OS, very trivial string error. Basically this step is try to generate a ground truth file. You can also manually generate it if you can't fix the error. Thanks.

**Comment by christofjan** [comment]
Hi sir,
          I resolved that error. i think the accuracy is not impressive compared with one you written in a report. could you please just give me some suggestion about following figures.. thank you.
accuracy for vgg16 = 63.7
prediction accuracy = 79.1 
accuracy_spatial =    36
accuracy_temporal =    43
accuracy_average =    39

**Comment by bryanyzhu** [comment]
What is the prediction accuracy? I thought vgg16 accuracy is the action prediction accuracy. 

And it is weird you couldn't achieve the performance as written in my paper, many people follow the procedure, and can get the numbers as mine. 

How do you get the images from the videos, cv2 or ffmpeg or sth else? I suspect your extracted images are quite different from mine. So when you use my trained model on your images, the performance is severely degraded. I suggest you use cv2 to decode your videos to frames.

**Comment by christofjan** [comment]
prediction accuracy =0. 79
accuracy = 0.6375
the procedure i followed was i generated frames using cv2 function and given frames path to .txt files and trained the model, which ran for 16000 iterations and it generated snapshot for every 500 iteration so i got 32 snapshots, i had given last snapshop (vgg16_end_iter_16000.caffemodel) path to 'model_file =' in demo_python.py file and ran the code, got  prediction accuracy (i.e0.79). finally i ran late_fusion.m file.
is there anything wrong in the steps i followed? please let me know if something is wrong.
here is the code written for extracting frames from videos. thank you..
#code for extracting videos
import os
import glob 
import sys
from pipes import quote
from multiprocessing import Pool, current_process

import argparse

def dump_frames(vid_path, out_path):
    import cv2
    video = cv2.VideoCapture(vid_path)
    vid_name = vid_path.split('.')[0]
    fcount = int(video.get(cv2.CAP_PROP_FRAME_COUNT))
    try:
        if not os.path.exists(out_path):
            os.makedirs(out_path)
    except OSError:
        pass
    file_list = []
    for i in xrange(fcount):
        ret, frame = video.read()
        # assert ret
        cv2.imwrite('{}/image_{:04d}.jpg'.format(out_path, i+1), frame)
        access_path = '{}/image_{:04d}.jpg'.format(vid_name, i+1)
        file_list.append(access_path)
    print '{} done'.format(vid_name)
    sys.stdout.flush()
    # return file_list
    return fcount

file_path = "/home/ch/Desktop/task/val_rgb_split1.txt"
training = "/home/ch/Desktop/task/training.txt"
validation = "/home/ch/Desktop/task/validation.txt"

with open(file_path) as read_file:
    txt = read_file.readlines() 

for line in txt:
        out_path = line.split()[0]
        video_path = "/home/ch/Downloads/videos/" + out_path.split('/')[-1] + ".avi"
        weight = line.split()[2]
        with open(validation, 'a+') as write_file:
            write_file.write(out_path + " " + str(dump_frames(video_path, out_path)) + " " + weight + "\n")

**Comment by bryanyzhu** [comment]
Your code seems to be correct, so I doubt you changed my code somewhere else, because many people have reproduced it. 

And sorry for my confusion, I still don't understand your situation here. If you get 

```
accuracy_spatial = 36
accuracy_temporal = 43
accuracy_average = 39
```

How can you get 63.7 or 79.1 accuracy? The `accuracy average` is your final number. 

Is 63.7 the accuracy during training or what? I just don't get it. 

Because in the matlab file, you should only get three outputs, accuracy_spatial for the spatial stream result, accuracy_temporal for the temporal stream result, and accuracy_average is the fusion result, which is the number I reported in the paper. Where did you get the 79.1? Thanks. 

I suggest just follow the readme, do the evaluation only. You should get the numbers correct, at least for the spatial stream.

**Comment by christofjan** [comment]
I only got three outputs for matlab file.
accuracy_spatial = 36
accuracy_temporal = 43
accuracy_average = 39
yes, i got 63.7 accuracy during training.
during evaluation i got 79.1
when i ran demo_hidden.py i got 79.1(prediction accuracy)

**Comment by bryanyzhu** [comment]
ok, I understand now. So for me, I got 70 accuracy during training. and demo_hidden.py give me about 84 accuracy. So the performance is ok, at least you are doing it right. 

So for 79.1, you are using my provided pre-trained model to get that number, right? If this is the case, the only problem will be your input. Your input is different from mine. Maybe you can check the number of frames per video. Compare it with my file list. They may have missing frames. 

If you are using your own model, then try different hyper-params and try to run several more times. You may get a higher number. 

But your matlab's result is so strange. Are you using my provided .mat file? 40 accuracy is so low. Somewhere must be wrong. But I couldn't reproduce your result on my end. So good luck, hope you can figure it out.

**Comment by christofjan** [comment]
yes i used pre-trained model!
i changed two lines in .mat files since my matlab version is not taking textread() function. I used textscan() instead. I suspect this might be the reason  to get that numbers.

original code 
groudTruth = textread(labelFile,'%s','delimiter','\n'); 

changed code
groudTruth = textscan(labelFile,'%s','delimiter','\n'); 
groudTruth = groudTruth{1};

i think i will try using textread() function in different version of matlab and let you know.. Thank you so much for your support. you patiently answered all my questions..

**Comment by christofjan** [comment]
Hi, i tried in different version of matlab. now i got correct numbers
accuracy_spatial =  0.8097
accuracy_temporal =  0.8406
accuracy_average =0.8900.
please don't misunderstand me, these questions may be the basic one, because i am completely new to this domain. i just want to learn more and experiment more on this. 
1.to predict actions on new videos where do i need to give the path of the video?
2.to test this model on real time action is there any existing code? or do i need to write a code for this?

thank you so much for your help.

**Comment by bryanyzhu** [comment]
That is great, glad you get the correct numbers. I am sorry I couldn't help more on that. 

And no problem at all, any questions are more than welcome. 

If you have a new video, you first use cv2 to decode the images into a folder, and give this path to this line [here](https://github.com/bryanyzhu/Hidden-Two-Stream/blob/master/models/ucf101_split1_unsup_end/eval_ucf101/demo_hidden.py#L52). You may need to modify demo_hidden.py a little bit, but not too much.

Basically, you need to delete these [lines](https://github.com/bryanyzhu/Hidden-Two-Stream/blob/master/models/ucf101_split1_unsup_end/eval_ucf101/demo_hidden.py#L32-L50) and the for loop, because you only need to test one video at a time. And then replace the input_video_dir. I think that is it.

**Comment by christofjan** [comment]
Thank you so much..

**Comment by christofjan** [comment]
Hi Sir, 
             I decoded images into folder "test" and given path to input_video_dir. I followed your suggestions given above,here  one think i didn't understand you told me to remove lines from 32-50 but without these lines i can't pass arguments to the HiddenTemporalPrediction() function which is assigned to the variable called spatial_prediction so I can't predict action label without doing this. right? so i just changed the input_video_dir and some lines in demo_hidden.py but I'm getting the following error. could you please tell me whether I am doing in right direction or not and please suggest me what that what are the changes should be made to the code..

here is the changed code and followed by error message..

`#!/usr/bin/env python

import os, sys
import numpy as np
import caffe
import math
import cv2
import scipy.io as sio
import h5py

from HiddenTemporalPrediction import HiddenTemporalPrediction

def softmax(x):
    y = [math.exp(k) for k in x]
    sum_y = math.fsum(y)
    z = [k/sum_y for k in y]

    return z

def main():

    # caffe init
    gpu_id = 0
    caffe.set_device(gpu_id)
    caffe.set_mode_gpu()

    # spatial prediction
    model_def_file = '../stack_motionnet_vgg16_deploy.prototxt'
    model_file = '../logs_end/ucf101_split1_vgg16_hidden.caffemodel'
    spatial_net = caffe.Net(model_def_file, model_file, caffe.TEST)
    start_frame = 0
    num_categories = 101
    feature_layer = 'fc8_vgg16'
    spatial_mean_file = './rgb_mean.mat'


    input_video_dir = "../test/"  # extracted video frames into folder test

    spatial_prediction = HiddenTemporalPrediction(
                input_video_dir,
                spatial_mean_file,
                spatial_net,
                num_categories,
                feature_layer,
		start_frame)
                   
  
    
    avg_spatial_pred_fc8 = np.mean(spatial_prediction, axis=1)
    avg_spatial_pred = np.asarray(softmax(avg_spatial_pred_fc8))
    predict_label = np.argmax(avg_spatial_pred)
    
    print input_video_dir
    print predict_label

    

if __name__ == "__main__":
    main()`


##error message

OpenCV Error: Assertion failed (ssize.width > 0 && ssize.height > 0) in resize, file /home/ch/Downloads/opencv-3.2.0/modules/imgproc/src/imgwarp.cpp, line 3492
Traceback (most recent call last):
  File "test.py", line 60, in <module>
    main()
  File "test.py", line 46, in main
    start_frame)
  File "/home/ch/Hidden-Two-Stream/models/ucf101_split1_unsup_end/eval_ucf101/HiddenTemporalPrediction.py", line 38, in HiddenTemporalPrediction
    img = cv2.resize(img, dims[1::-1])
cv2.error: /home/ch/Downloads/opencv-3.2.0/modules/imgproc/src/imgwarp.cpp:3492: error: (-215) ssize.width > 0 && ssize.height > 0 in function resize

**Comment by bryanyzhu** [comment]
Yes, you are right. Line 32 to 50 is useful, what I mean is you don't need the for loop. Your code looks good to me.

The error is pretty clear. It is in the resize function. There are two possibilities.
1. The images are not read correctly. Check the result of this [line](https://github.com/bryanyzhu/Hidden-Two-Stream/blob/master/models/ucf101_split1_unsup_end/eval_ucf101/HiddenTemporalPrediction.py#L37). Maybe it is a `None`. If this is the case, check the image loading procedure [here](https://github.com/bryanyzhu/Hidden-Two-Stream/blob/master/models/ucf101_split1_unsup_end/eval_ucf101/HiddenTemporalPrediction.py#L21-L22). Maybe your images do not end in `.jpg`.

2. Your image size is too small for resize and crop. But I think it is the first reason.

**Comment by christofjan** [comment]
Hi, i resolved that error. but i'm getting wrong predictions. I guess i have to give more precise action videos as input. I will try to get it right. once again thank you so much for the help..

**Comment by bryanyzhu** [comment]
No problem, glad I can help. I will close the issue for now. If you have other questions, feel free to open a new one.
